poradnia/letters/templatetags/parse_markdown.py

```diff
@@ -0,0 +1,21 @@
+from django import template
+from django.template.defaultfilters import stringfilter
+from django.utils.safestring import mark_safe
+
+import markdown as md
+from bleach.sanitizer import Cleaner
+
+
+cleaner = Cleaner(
+tags=['a', 'abbr', 'acronym', 'b', 'blockquote', 'code', 'em', 'i', 'li', 'ol', 'strong', 'ul', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'p', 'hr', 'pre', 'img'],
+attributes={'a': ['href', 'title'],'img': ['alt', 'src', 'title'] ,'abbr': ['title'], 'acronym': ['title']})
+
+register = template.Library()
+
+
+@register.filter()
+@stringfilter
+def markdown(value):
+    markdown = md.markdown(value, extensions=['markdown.extensions.fenced_code', 'sane_lists'])
+    sanitized = cleaner.clean(markdown)

```

Comment:
Możemy po prostu nie obsługiwać HTML w Markdown, zamiast rysować filtrowaniem?

Reply:
To znaczy tak - jeśli chodzi o HTML w samej treści pola tekstowego to on jest nieobsługiwany, zostanie przerobiony na zwykły test.

Natomiast sam parser zwraca już w templatce HTML i dla bezpieczeństwa filtruję to wyjście pozwalając tylko na HTML, który zawiera jedynie znaczniki definiowane przez Markdown.

Nie jestem pewien czy odpowiedziałem na twoje pytanie :) W razie potrzeby rozwiń proszę.

Reply:
Poprawnym elementem Markdown jest kod HTML na wejściu. Jednak kodu HTML na wejściu nie potrzebujemy obsługiwać, więc w mojej ocenie parser Markdown powinien escapować cały kod HTML, który napotka. Na wyjściu otrzymamy zatem tylko kontrolowany przez nas zakres znaczników HTML wynikający tylko i wyłącznie z zakresu wspieranej składni Markdown. Powinniśmy obsługiwać tylko składnie Markdown, która jest bezpieczna. Wtedy bleach nam nie jest potrzebny, a transformacji jest mniej. 

Nie znam bibliotek Python dla Markdown, ale zobacz na parametr "html" w markdown-it ( https://github.com/markdown-it/markdown-it/blob/master/README.md ).

Reply:
Ok, teraz się rozumiemy.
W pełni się zgadzam i dokładnie jest to realizowane teraz tak jak mówisz.
To jest HTML na wejściu jest escapowany w całości domyślnie.

Natomiast nie zabezpiecza to strony przed atakami XSS, ponieważ można wykorzystać do tego sam Markdown, dlatego sanityzuję również wyjście. Nie zadziała tutaj żaden filtr na wejściu ponieważ mieszamy składnię HTML/JS/Markdown, którego filtr nie rozpozna. Dlatego dla 100% bezpieczeństwa należy sprawdzać output, który już jest w 100% do zweryfikowania.

Przykład:
```markdown
 [some text](javascript:alert('xss'))
```
zwraca złośliwy HTML:
```html 
<a href="javascript:alert('xss')">some text</a>
```
Przykład 2:
```markdown
> hello <a name="n"
> href="javascript:alert('xss')">*you*</a>
```
HTML
```html
<blockquote>
 <p>hello <a name="n"
 href="javascript:alert('xss')"><em>you</em></a></p>
</blockquote>
```

Reply:
W zakresie pierwszego przykładu Bleach przed tym nie zabezpiecza i myślę, że to właśnie Markdown ma większość możliwość weryfikowania co jest linkiem niż Bleach.

W zakresie drugiego przykładu wydaje mi się, że powinien on być przez parser Markdown zwrócony jako:
> hello &lt;a name=&quot;n&quot;
> href=&quot;javascript:alert('xss')&quot;&gt;*you*&lt;/a&gt;&lt;/p&gt;

W takim przypadku jest bezpieczny.

Reply:
Dotąd postrzegałem jako coś co zabezpiecza przed XSS, bo ma prostszą składnie, stąd poddaje w wątpliwość Bleach.

W markdown-it na https://markdown-it.github.io/ dla:
```
[some text](javascript:alert('xss'))

> hello <a name="n"
> href="javascript:alert('xss')">*you*</a>
```
otrzymuje się:
```
<p>[some text](javascript:alert(‘xss’))</p>
<blockquote>
<p>hello &lt;a name=“n”
href=“javascript:alert(‘xss’)”&gt;<em>you</em>&lt;/a&gt;</p>
</blockquote>
```

To wygląda bezpiecznie i nie wymaga dodatkowo filtrowania.

Reply:
Rzeczywiście w markdown-it escapują również HTML na wyjściu w standardzie.
W takim razie mogę śmiało wymienić tę bibliotekę i odpadnie nam jedno dependency. 
Dzięki. Zcommituja coś wieczorem :)

Reply:
markdown-it jest dla JavaScriptu :) Proponuje coś poszukać dla Pythona zbliżonego, bo mniejsze ryzyko ma po prostu wbudowanie tego w parser Markdown.

Reply:
Jest port do Pythona, także zaraz zobaczymy w akcji.
https://pypi.org/project/markdown-it-py/

Reply:
Podoba mi się biblioteka, zwłaszcza ta strona dokumentacji mnie przekonuje https://markdown-it-py.readthedocs.io/en/latest/other.html . Od strony funkcjonalności nie mamy wymagań, ale taki opis w dokumentacji brzmi, że nie będzie biblioteka sprawiać problemów w przyszłości.

Reply:
Niestety `markdown-it-py` ma spore problemy z poprawnym formatowaniem. Natomiast znalazłem alternatywę w postaci biblioteki [Mistune](https://mistune.readthedocs.io/en/latest/guide.html). Ona również ma już zaimplementowane zabezpieczenia i HTML na wejściu jest escape'owany a wyjście sanityzowane. Sprawdza się przy XSSach o których pisaliśmy i jednocześnie nie rozwala formatowania, także myślę, że zostałbym przy niej.

Reply:
Możesz podać przykłady, gdy `markdown-it-py` nie radził sobie z formatowaniem?

Reply:
Coś jest nie tak z renderowaniem generalnie. Przy listach całkowicie wyrzuca błędy.
![List error](https://user-images.githubusercontent.com/23581770/95504080-88dad600-09ac-11eb-8579-6d0a633e52b1.png)

Po wyłączeniu list markdown się renderuje z wieloma błędami.
![Porównanie tekstu demo markdown-it i rezultatów](https://user-images.githubusercontent.com/23581770/95503954-621c9f80-09ac-11eb-8932-7b7495508a6f.png)

Zgłosiłem buga, być może ktoś z ich strony będzie mógł to sprawdzić i da znać co może byc nie tak.

Niestety w Mistune również przechodzi jeden typ ataków XSS (z użyciem img:onerror).

Sugeruję w tym przypadku by jednak wrócić do używania Bleacha, który nie przepuści tagów innych niż dozwolone.

Co sądzisz?.

Reply:
Skoro nie mamy alternatywy to użyjmy Bleach i odnotujmy, że można by go w przyszłości usunąć, bo teoretycznie jest zbędny.